README
Module    Documentation ~ Compiled ~ Examples Install ~ Regen Docs

apache          *            *             *                *
aspell          *            *             *                *
crypt           *            *             *                *
curl            *            *             *                *
curses          *            *             *                *
dbi             *            *             *                *
embfer          *            *             *                *
gettext         *            *             *                *
image           *            *             *                *
mail
MHash           *            *             *                *
serial          *            *             *                *
template        *            *             *                *
ui              X            X
uxml            X            X                              
xslt            *            *             *                *
zip             *            *             *                *


dbi/ChangeLog
-e --------------------------------------------------------------------------------
08.07.2008 11:06 (BST), chris

 * A function for queryUpdateInsert
 * A new part of the module dbi_sql.fe which provides code based SQL query generation

-e --------------------------------------------------------------------------------
27.06.2008 10:16 (BST), chris

 * Query tracking with timestamps

--------------------------------------------------------------------------------
22.09.2006 14:54 (CEST), cms

  * added an overloaded connect function that takes the port number as a
    parameter.

--------------------------------------------------------------------------------
13.03.2006 23:41 (UTC), cms

  * Added some metadata functions. 
   
 	 columnName( number column )
	 columnLength( number column )   	 
	 columnIsNull( number column )
	 columns( )

--------------------------------------------------------------------------------
10.01.2006 22:02 (UTC), cms

   * Make dbi free its malloc'ed memory as soon as its not needed any more. 
     Before all the memory was free'ed from Dbis destructor which was a bit
     devastating for daemons and such which might run for a very long time and
     alloc alot of DbiResults.

   * Some cleanups and documentation.

--------------------------------------------------------------------------------
08.01.2006 12:30 (GMT), chris

 * Make stored procedures much better

--------------------------------------------------------------------------------
23.11.2005 13:24 (GMT), chris

  * You can now call stored procedures by calling SP_NAMEOFPROC() on the 
    dbi object

--------------------------------------------------------------------------------
16.12.2004 10:17 (GMT), chrison

  * Change to the build script to use prefix

--------------------------------------------------------------------------------
07.09.2004 20:19, cms
	* removed the free() on error strings.
          libdbi now take care of free()'ing the error strings.

--------------------------------------------------------------------------------
21.08.2004 14:10 (CEST), cms

	* Renamed function names from function_name style to functionName style
          as this is the standard way for ferite....

--------------------------------------------------------------------------------
25.03.2004 16:33 (CET), cms

 * interfaces made "protocol" a reserved word.


--------------------------------------------------------------------------------
18.12.2003 13:34 (UTC), cms

  * Some cleanups

--------------------------------------------------------------------------------
15.12.2003 10:27 (CET), cms

 * fix for a segfault

--------------------------------------------------------------------------------
18.02.2003 12:41 (CET), cms

 * fix for function parameter count.

--------------------------------------------------------------------------------
07.08.2002 12:28 (CEST), cms

  * Handle Integers correctly.



--------------------------------------------------------------------------------
03.08.2002 18:26 (CEST), cms

  * Sync DBI build process with ferite's (Solaris safe) and cleanups



--------------------------------------------------------------------------------
17.07.2002 14:16 (CEST), cms

 * make dbi duplicate script safe.



--------------------------------------------------------------------------------
10.07.2002 16:25 (BST), chris

  * fix to some comments

--------------------------------------------------------------------------------
05.07.2002 19:13 (CEST), cms

* bug fix, cleanup and make query_binary() return NULL.



--------------------------------------------------------------------------------
05.07.2002 17:27 (BST), chris

  * Return a null object if we are not connected to the database

--------------------------------------------------------------------------------
14.06.2002 18:47 (CEST), cms

** Warning This commit will break all code using dbi. Warning **

* Cleanup.
* disconnect is back. 
* Removed connect();

  The constructor now does the connection. 
 
 	uses  "dbi","console";
	object dbh = new Dbi(protocal, username, ...); 
	if(dbh == null)
		Console.println("Danger ...");


--------------------------------------------------------------------------------
14.06.2002 18:18 (CEST), cms

seq_curr_val();
seq_next_val();



--------------------------------------------------------------------------------
14.06.2002 11:04 (CEST), sem

* Removed the Self* macro to make debugging easier
* Removed Destruct for dbi_rresult
* Removed the dbi_result::finished() and dbi::disconnect()
* Took away module-deinit{} that contained dbi_shutdown()
* All the problems was due to dbi_shutdown() running before GC

--------------------------------------------------------------------------------
31.05.2002 17:04 (CEST), cms

* removed query_placholders(). Query now has this functionality. 
	*** Warning this will probably break stuff ***

--------------------------------------------------------------------------------
31.05.2002 14:12 (CEST), cms

* first attempt om emulated placeholders 
  called query_placeholders untill its tested enough to become query.

--------------------------------------------------------------------------------
31.05.2002 11:21 (CEST), andreas

Added function driver_name which retreives the name of the driver
mysql for MySQL
pgsql for PostGreSQL

--------------------------------------------------------------------------------
27.05.2002 22:16 (CEST), sem

* Added all_records() to return an array of arrays

--------------------------------------------------------------------------------
27.05.2002 21:59 (CEST), sem

* You can return a database row as an object, without methods that is

--------------------------------------------------------------------------------
27.05.2002 15:26 (CEST), sem

*updated test script to use new "sys" module, for use with cvs ferite




crypt/crypt/base.c
#include <string.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <ferite.h>
#include <math.h>


unsigned char alphabet[64] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
static const char pr2six[256] = {
    /* ASCII table */
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63,
    52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64,
    64,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
    15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64,
    64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
    41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
};

int base64decode_len(const char *bufcoded) {
    int nbytesdecoded;
    register const char *bufin;
    register int nprbytes;
    bufin = bufcoded;
    while (pr2six[*bufin++] <= 63);
    nprbytes = (bufin - (const char *) bufcoded) - 1;
    nbytesdecoded = ((nprbytes + 3) / 4) * 3;
    return nbytesdecoded + 1;
}

int base64decode(char *bufplain, const char *bufcoded) {
    int nbytesdecoded;
    register const char* bufin;
    register char *bufout;
    register int nprbytes;

    bufin = bufcoded;
    while (pr2six[*(bufin++)] <= 63);
    nprbytes = (bufin - (const char *) bufcoded) - 1;
    nbytesdecoded = ((nprbytes + 3) / 4) * 3;

    bufout = (char *) bufplain;
    bufin = (const char *) bufcoded;

    while (nprbytes > 4) {
        *bufout++ = (pr2six[(unsigned char) *bufin] << 2 | pr2six[(unsigned char)bufin[1]] >> 4);
        *bufout++ = (pr2six[(unsigned char)bufin[1]] << 4 | pr2six[(unsigned char)bufin[2]] >> 2);
        *bufout++ = (pr2six[(unsigned char)bufin[2]] << 6 | pr2six[(unsigned char)bufin[3]]);
        bufin += 4;
        nprbytes -= 4;
    }

    /* Note: (nprbytes == 1) would be an error, so just ingore that case */
    if (nprbytes > 1)
        *bufout++ = (pr2six[*bufin] << 2 | pr2six[bufin[1]] >> 4);
    if (nprbytes > 2)
        *bufout++ = (pr2six[bufin[1]] << 4 | pr2six[bufin[2]] >> 2);
    if (nprbytes > 3)
        *bufout++ = (pr2six[bufin[2]] << 6 | pr2six[bufin[3]]);

    *bufout++ = '\0';
    nbytesdecoded -= (4 - nprbytes) & 3;
    return nbytesdecoded;
}


char* base64encode(unsigned char *ip, int insize)
{
    int  bits = 0;
    int in = 0, out = 0, rem = 0,i,index;
    unsigned char *op,*start,*test;
    if( (start = op = fmalloc_ngc(4*(insize/3) + 4*((insize%3) > 0) + 1 )) == NULL )
	    return NULL;
    while ( in++ < insize ) {
	    rem++;
	    bits += *ip++ << 8*(3-rem);
	    if ( rem == 3 ) {
		    for(;rem>=0;rem--) {
		      index = (bits >> 6*rem) & 0x3f;
		      *op++ =  alphabet[index];
		      
		    }
		    bits = rem = 0;
	    }
    }
    if ( rem )
	    for(i= 3 ; i>=0 ;i--)
		    if( (i > 1 ) || ( i==1 && rem ==2))
			    *op++ = alphabet[(bits >> 6*i) & 0x3f];
		    else
			    *op++ = '=';
    
    *op = '\0';
   
    return start;
}

